File: August 2021/August_05.java
class Solution {
    private int helper(int[] piles, int s, int e, int p, Integer[][][] dp){
        if(s > e)
            return 0;
        if(dp[s][e][p] != null)
            return dp[s][e][p];
        int x = helper(piles, s + 1, e, 1 - p, dp);
        int y = helper(piles, s, e - 1, 1 - p, dp);
        if(p == 0){
            dp[s][e][p] = Math.max(piles[s] + x, piles[e] + y);
            return dp[s][e][p];
        }
            
        else{
            dp[s][e][p] = Math.min(x - piles[s], y - piles[e]);
            return dp[s][e][p];
        }
    }
    public boolean stoneGame(int[] piles) {
        return helper(piles, 0, piles.length - 1, 0, new Integer[piles.length][piles.length][2]) > 0;
    }
}

File: August 2022/August_05.cpp
class Solution {
public:
    int helper(vector<int> &nums, int id, int target, vector<vector<int>> &dp){
        if(target < 0)
            return 0;
        if(id == nums.size()){
            if(target == 0)
                return 1;
            return 0;
        }
        if(dp[id][target] != -1)
            return dp[id][target];
        return dp[id][target] = helper(nums, 0, target - nums[id], dp) + helper(nums, id + 1, target, dp);
    }
    
    int combinationSum4(vector<int>& nums, int target) {
        int n = nums.size();
        vector<vector<int>> dp(n, vector<int>(target +  1, -1));
        return helper(nums, 0, target, dp);
    }
};
